### PR TITLE
Add "close button"

### PR DESCRIPTION
**Pull request by sh19910711**
The main changes of this pull request are the followings:

* ~~Change styles of the resizer element~~
* Change the positioning of the console elements
* Add "Close button"
* ~~Add "Minimize button"~~

> Well, the absolute positioning is to add more elements like the resizer element. For example, I think to add "minimize button", "close button" and "setting button" to the console, and the absolute positioning is able to show multiple elements like that to the top of the console element.

The quoted text is from PR #139.

#### GIF

gif

### CODE REVIEW

**Comment by sh19910711** [comment]
I have pushed commits as the first version which contains wip-commits for code reviews. Thanks.

**Comment by gsamokovarov** [comment]
Because we share the browser space with user content, I think we shouldn't waste that much real estate with the resize bar. We can keep it as is, and provide just the close button. If people want the console out of the way, they can just resize it down. Further more, we can probably drop the outline of the close button and show it only when the mouse hovers the console (so again, we save the real estate).

We also should be able to opt-out of this behaviour for the browser extensions.

**Comment by sh19910711** [comment]
@gsamokovarov Thanks for the reply.

> Because we share the browser space with user content, I think we shouldn't waste that much real estate with the resize bar.

I understand that.

> We can keep it as is, and provide just the close button. If people want the console out of the way, they can just resize it down.

And I also agree with that.

> Further more, we can probably drop the outline of the close button and show it only when the mouse hovers the console (so again, we save the real estate).

Uhm, I'm not quite sure of that. The next link shows my interpretation: http://jsfiddle.net/bmvfLkjq/4/

Let me know if it is incorrect.

> We also should be able to opt-out of this behaviour for the browser extensions.

Yeah, I think to add a config of the REPLConsole constructor.

Thank you.

**Comment by gsamokovarov** [comment]
> Uhm, I'm not quite sure of that. The next link shows my interpretation: http://jsfiddle.net/bmvfLkjq/4/

My thought was to keep the resize bar transparent as before. We may not have the close button attached to it (at least visually), having the close boutton couple of pixels down to the right of it may cut it. Something like this ASCII:

```
+---------------------------+  <- You can assume that this line is
|                         + |     the resize bar :-)
|                           |
|                           |
|                           |
|                           |
+---------------------------+
```

> Yeah, I think to add a config of the REPLConsole constructor.

Exactly.

**Comment by sh19910711** [comment]
@gsamokovarov Thanks for giving an example :-) OK, I will think of it.

**Comment by sh19910711** [comment]
I have pushed new version with the following changes:

* Remove the minimize button
* The `shiftConsoleActions()` is for the fixed positioning

@gsamokovarov Hi, could you review the code again?

And here is a gif:

gif

Thank you.

File: lib/web_console/templates/style.css.erb [hunk 1]
```diff
@@ -1,9 +1,22 @@
...
+.console { position: fixed; left: 0; bottom: 0; width: 100%; height: 148px; padding: 0; margin: 0; background: none repeat scroll 0% 0% #333; z-index: 9999; }
+.console .console-outer { overflow: auto; padding-top: 4px; }
+.console .console-inner { font-family: monospace; font-size: 11px; width: 100%; height: 100%; overflow: none; background: #333; }
+.console .console-prompt-box { color: #FFF; }
+.console .console-message { color: #1AD027; margin: 0; border: 0; white-space: pre-wrap; background-color: #333; padding: 0; }
+.console .console-focus .console-cursor { background: #FEFEFE; color: #333; font-weight: bold; }
+.console .resizer { background: #333; width: 100%; height: 4px; cursor: ns-resize; }
+.console .console-actions { padding-right: 3px; }
+.console .console-actions .button { float: left; }
+.console .button { cursor: pointer; border-radius: 1px; font-family: monospace; font-size: 9px; width: 10px; height: 10px; line-height: 10px; text-align: center; color: #ccc; }
```
Comment (gsamokovarov) [inline lib/web_console/templates/style.css.erb]: Can we make the close button a bit bigger? Say `font-size: 13px, width: 14px; height: 14px; line-height: 14px;`, I tested on my machine and it looks reasonable.

**Comment by gsamokovarov** [comment]
Thanks @sh19910711, this looks great! Now that we have templates tests infrastructure, we can try to write a regression test for this.

**Comment by sh19910711** [comment]
> Now that we have templates tests infrastructure, we can try to write a regression test for this.

OK, I will try it.

Thanks for taking time to check :)

File: test/templates/spec/repl_console_spec.js [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+          this.replConsole = replConsole;
+        });
+
+        context("click close button", function() {
+          beforeEach(function() {
+            var closeButton = this.elm.getElementsByClassName("close-button")[0];
+            SpecHelper.triggerEvent(closeButton, "click");
+          });
+          it("should be removed from the parent node", function() {
+            assert.ok(document.getElementById(this.elmId) === null);
```
Comment (gsamokovarov) [inline test/templates/spec/repl_console_spec.js]: Isn't there assert.nil or something in the lines of?

File: test/templates/spec/repl_console_spec.js [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+  describe("Console Actions", function() {
+    describe("Close Action", function() {
+      beforeEach(function() {
+        this.elm = document.createElement("div");
+        this.elmId = this.elm.id = SpecHelper.randomString();
+        stageElement.appendChild(this.elm);
+      });
+
+      context("install console", function() {
```
Comment (gsamokovarov) [inline test/templates/spec/repl_console_spec.js]: Do we need that context? We already test the install console, no need to keep it here just to borrow its beforeEach functions.

File: test/templates/spec/repl_console_spec.js [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+      it("should have a resizer", function() {
+        assert.equal(this.elm.getElementsByClassName("resizer").length, 1);
+      });
+      it("should have a close button", function() {
+        assert.equal(this.elm.getElementsByClassName("close-button").length, 1);
+      });
+    });
+  });
+
+  describe("Console Actions", function() {
```
Comment (gsamokovarov) [inline test/templates/spec/repl_console_spec.js]: Can we lowercase those as the rest of the test descriptions?

File: test/templates/spec/repl_console_spec.js (hunk 2 above)
Comment (sh19910711) [inline test/templates/spec/repl_console_spec.js]: Yes, I found `assert.isNull` :)

http://chaijs.com/api/assert/#isNull-section

File: test/templates/spec/spec_helper.js [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+      var event = document.createEvent("MouseEvents");
+      event.initEvent(eventName, true, true); // type, bubbles, cancelable
+      el.dispatchEvent(event);
+    },
+    randomString: function() {
+      Math.random().toString(36).substring(2);
+    }
+  };
+
+  var stageElement = document.createElement("div");
```
Comment (gsamokovarov) [inline test/templates/spec/spec_helper.js]: Is that something that is needed for every test we have? We may make it callable and include it only where needed.

File: test/templates/spec/repl_console_spec.js (hunk 2 above)
Comment (gsamokovarov) [inline test/templates/spec/repl_console_spec.js]: Let's use it then :)

**Comment by sh19910711** [comment]
Updated (and here is a [diff](https://github.com/sh19910711/web-console/compare/7ad6d18f501a5ea9c729f3230fac89e6a6333676...7d2e8d8d7f0513f481f22ab85d2711a76ce2ae72)). The `SpecHelper.prepareStageElement()` is to do as its name implies.

**Pull request merged**